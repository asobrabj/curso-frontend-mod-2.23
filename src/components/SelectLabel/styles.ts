import styled from 'styled-components'

interface ISelectLabelStyle {
  $fs?: 'xsmall' | 'small' | 'medium' | 'larger' | string
  $fStyle?: 'normal' | 'italic'
  $fw?: 'medium' | 'larger'
  $fFamily?: string
  $tAlign?: 'start' | 'end' | 'justify' | 'center'
  $tTransform?: 'none' | 'capitalize' | 'uppercase' | 'lowercase'
  $color?: string
  $bgColor?: string
  $pd?: 'xsmall' | 'small' | 'medium' | 'larger' | 'xlarger' | 'none'
  $pdTop?: string
  $pdBottom?: string
  $pdRigth?: string
  $pdLeft?: string
  $mg?: 'xsmall' | 'small' | 'medium' | 'larger' | 'xlarger' | 'none'
  $mgTop?: string
  $mgBottom?: string
  $mgRigth?: string
  $mgLeft?: string
  $width?: string
  $height?: string
  $visible: boolean
  $bdColor?: string
}

interface Defaults {
  [key: string]: string
}

const fontSize = (fs: string): string => {
  const defaults: Defaults = {
    xsmall: '14',
    small: '16',
    medium: '24',
    larger: '32',
  }
  const value = defaults[`${fs}`] || fs
  return `${value}px`
}

const fontWheigt = (fw: string): string => {
  const defaults: Defaults = {
    medium: '400',
    larger: 'bold',
  }
  const value = defaults[`${fw}`]
  return value
}

const paddingFn = (pd: string): string => {
  const defaults: Defaults = {
    xsmall: '5',
    small: '8',
    medium: '16',
    larger: '24',
    xlarger: '32',
  }
  const value = defaults[`${pd}`] ? `${defaults[`${pd}`]}px` : ''
  return value
}

const marginFn = (mg: string): string => {
  const defaults: Defaults = {
    xsmall: '5',
    small: '8',
    medium: '16',
    larger: '24',
    xlarger: '32',
  }
  const value = defaults[`${mg}`] ? `${defaults[`${mg}`]}px` : ''
  return value
}

export const Label = styled.label`
  position: absolute;
  top: -12px;
  left: 12px;
  background-color: '';

  span {
    font-size: 14px;
    color: red;
    margin-right: 3px;
  }
`

export const Select = styled.div<ISelectLabelStyle>`
  font-style: ${(props) => props.$fStyle || ''};
  font-weight: ${(props) => (props.$fw ? fontWheigt(props.$fw) : '')};
  font-family: ${(props) => props.$fFamily || ''};
  text-align: ${(props) => props.$tAlign || ''};
  text-transform: ${(props) => props.$tTransform || ''};
  color: ${(props) => props.$color || ''};
  background-color: ${(props) => props.$bgColor || ''};
  padding: ${(props) => (props.$pd ? paddingFn(props.$pd) : '')};
  padding-top: ${(props) => (props.$pdTop ? props.$pdTop + 'px' : '')};
  padding-left: ${(props) => (props.$pdLeft ? props.$pdLeft + 'px' : '')};
  padding-right: ${(props) => (props.$pdRigth ? props.$pdRigth + 'px' : '')};
  padding-bottom: ${(props) => (props.$pdBottom ? props.$pdBottom + 'px' : '')};
  margin: ${(props) => (props.$mg ? marginFn(props.$mg) : '')};
  margin-top: ${(props) => (props.$mgTop ? props.$mgTop + 'px' : '')};
  margin-left: ${(props) => (props.$mgLeft ? props.$mgLeft + 'px' : '')};
  margin-right: ${(props) => (props.$mgRigth ? props.$mgRigth + 'px' : '')};
  margin-bottom: ${(props) => (props.$mgBottom ? props.$mgBottom + 'px' : '')};
  width: ${(props) => (props.$width ? props.$width + 'px' : '')};
  height: ${(props) => (props.$height ? props.$height + 'px' : '')};
  border: 1px solid #00abff;
  border-color: ${(props) => props.$bdColor || ''};
  position: relative;

  &.primary {
    width: ${(props) => (props.$width ? props.$width + 'px' : '200px')};
    padding: ${(props) => (props.$pd ? paddingFn(props.$pd) : '8px')};
    border-radius: 8px;
    border-color: ${(props) => props.$bdColor || ''};
  }

  &.secondary {
    width: ${(props) => (props.$width ? props.$width + 'px' : '200px')};
    padding: ${(props) => (props.$pd ? paddingFn(props.$pd) : '8px')};
    border-radius: 8px 0;
    border-color: ${(props) => props.$bdColor || ''};
  }

  &.error {
    border: 1px solid #ff0000;
    color: ${(props) => props.$color || '#ff0000'};
    &:focus {
      border: 2px solid #ff0000;
    }
  }
  &:active {
    border-color: ${(props) => props.$bdColor || ''};
  }

  & select {
    display: none;
  }

  &.custom-select {
    position: relative;
    display: inline-block;
    width: 100%;
  }

  .custom-select-trigger {
    width: 100%;
    cursor: pointer;
    height: 20px;
  }

  .custom-options {
    background-color: '';
    display: ${(props) => (props.$visible ? 'block' : 'none')};
    position: absolute;
    top: 100%;
    left: 0px;
    right: 0;
    background-color: #fff;
    border: 1px solid #00abff;
    max-width: ${(props) => (props.$width ? props.$width + 'px' : '215px')};
    width: 100%;
    max-height: 114px;
    overflow-y: auto;
    z-index: 1;
    font-size: ${(props) => (props.$fs ? fontSize(props.$fs) : '')};

    &.primary {
      border-radius: 8px;
      border-color: ${(props) => props.$bdColor || ''};
    }
    &.secondary {
      border-radius: 8px 0;
      border-color: ${(props) => props.$bdColor || ''};
    }
  }

  .custom-option {
    padding: 10px;
    cursor: pointer;
    font-size: ${(props) => (props.$fs ? fontSize(props.$fs) : '')};
  }

  .custom-option:hover {
    background-color: '';
    color: ${(props) => props.$color || '#00abff'};
  }
`
