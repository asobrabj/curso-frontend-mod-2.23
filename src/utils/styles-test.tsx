import { render } from '@testing-library/react'
import { ThemeProvider } from 'styled-components'
import theme from '../styles/darkTheme'

export const renderTheme = (Children: React.ReactNode) => {
  return render(<ThemeProvider theme={theme}> {Children} </ThemeProvider>)
}
