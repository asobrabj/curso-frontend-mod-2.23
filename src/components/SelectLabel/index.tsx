import { useState } from 'react'
import * as S from './styles'

export interface ISelectLabel {
  listOptions: Array<string>
  fs?: 'xsmall' | 'small' | 'medium' | 'larger' | string
  fStyle?: 'normal' | 'italic'
  fw?: 'medium' | 'larger'
  fFamily?: string
  tAlign?: 'start' | 'end' | 'justify' | 'center'
  tTransform?: 'none' | 'capitalize' | 'uppercase' | 'lowercase'
  color?: string
  bgColor?: string
  pd?: 'xsmall' | 'small' | 'medium' | 'larger' | 'xlarger' | 'none'
  pdTop?: string
  pdBottom?: string
  pdRigth?: string
  pdLeft?: string
  mg?: 'xsmall' | 'small' | 'medium' | 'larger' | 'xlarger' | 'none'
  mgTop?: string
  mgBottom?: string
  mgRigth?: string
  mgLeft?: string
  className?: string
  width?: string
  height?: string
  placeholder?: string
  bdColor?: string
  required: boolean
  label: string
  id: string
  onChange?: (e: React.ChangeEvent<HTMLSelectElement>) => void
  value?: string
}

export default function SelectLabel({
  listOptions,
  fs,
  fStyle,
  fw,
  fFamily,
  tAlign,
  tTransform,
  color,
  bgColor,
  pd,
  pdTop,
  pdRigth,
  pdLeft,
  pdBottom,
  mg,
  mgTop,
  mgRigth,
  mgLeft,
  mgBottom,
  className,
  width,
  height,
  placeholder,
  bdColor,
  required = false,
  label,
  id,
  onChange,
  value,
}: ISelectLabel) {
  const [visible, setVisible] = useState<boolean>(false)
  const [selectedValue, setSelectedValue] = useState<string | undefined>(
    placeholder || ''
  )

  const changeVisible = () => {
    setVisible((v) => !v)
  }
  const handleOptionClick = (value: string) => {
    setSelectedValue(value)
    setVisible(false)
    if (onChange) {
      onChange({
        target: {
          value,
          name: id,
        },
      } as React.ChangeEvent<HTMLSelectElement>)
    }
  }

  return (
    <S.Select
      className={`primary ${className}`}
      $fs={fs}
      $fw={fw}
      $fStyle={fStyle}
      $fFamily={fFamily}
      $tAlign={tAlign}
      $tTransform={tTransform}
      $color={color}
      $bgColor={bgColor}
      $pd={pd}
      $pdTop={pdTop}
      $pdRigth={pdRigth}
      $pdLeft={pdLeft}
      $pdBottom={pdBottom}
      $mg={mg}
      $mgTop={mgTop}
      $mgRigth={mgRigth}
      $mgLeft={mgLeft}
      $mgBottom={mgBottom}
      $width={width}
      $height={height}
      $visible={visible}
      $bdColor={bdColor}
    >
      <div className="custom-select">
        <S.Label htmlFor={id}>
          {' '}
          {required && <span>*</span>}
          {label}
        </S.Label>
        <div
          className={`custom-select-trigger primary ${className}`}
          onClick={changeVisible}
          id={id}
          data-testid={`select-${id}`}
        >
          {selectedValue}
        </div>
        <div className={`custom-options primary ${className}`}>
          {listOptions &&
            listOptions.map((option) => (
              <div
                key={option}
                className="custom-option"
                data-value={option}
                onClick={() => handleOptionClick(option)}
                data-testid="options"
              >
                {option}
              </div>
            ))}
        </div>
        <select
          id="real-select"
          required={required}
          value={value}
          onChange={onChange}
        >
          {listOptions.map((option) => (
            <option key={option} value={option}>
              {option}
            </option>
          ))}
        </select>
      </div>
    </S.Select>
  )
}
