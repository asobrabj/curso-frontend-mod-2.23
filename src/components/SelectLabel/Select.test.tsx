import { act, fireEvent, screen, waitFor } from '@testing-library/react'
import SelectLabel, { ISelectLabel } from '.'
import { renderTheme } from '../../utils/styles-test'

describe('SelectLabel Component', () => {
  const mockOptions: string[] = ['Option 1', 'Option 2', 'Option 3']

  const renderSelectLabel = (props: Partial<ISelectLabel> = {}) => {
    const defaultProps: ISelectLabel = {
      listOptions: mockOptions,
      required: false,
      label: 'Select Label',
      id: 'select-label',
      onChange: jest.fn(),
      value: '',
    }

    const mergedProps = { ...defaultProps, ...props }

    return renderTheme(<SelectLabel {...mergedProps} />)
  }

  it('deve renderizar corretamente', () => {
    renderSelectLabel()

    const selectLabelElement = screen.getByTestId('select-select-label')
    expect(selectLabelElement).toBeInTheDocument()
  })

  it('deve exibir as opções ao clicar no trigger', () => {
    renderSelectLabel()

    const triggerElement = screen.getByText('Select Label')
    act(() => {
      fireEvent.click(triggerElement)
    })

    waitFor(() => {
      const option1 = screen.getByText('Option 1')
      const option2 = screen.getByText('Option 2')
      const option3 = screen.getByText('Option 3')
      expect(option1).toBeInTheDocument()
      expect(option2).toBeInTheDocument()
      expect(option3).toBeInTheDocument()
    })
  })
  it('snapshot', () => {
    const { container } = renderTheme(<SelectLabel listOptions={mockOptions} required={false} label={''} id={''} />)
    expect(container).toMatchSnapshot()
  })
})
