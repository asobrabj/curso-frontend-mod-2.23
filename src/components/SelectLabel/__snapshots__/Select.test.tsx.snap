// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectLabel Component snapshot 1`] = `
.c1 {
  position: absolute;
  top: -12px;
  left: 12px;
  background-color: '';
}

.c1 span {
  font-size: 14px;
  color: red;
  margin-right: 3px;
}

.c0 {
  border: 1px solid #00abff;
  position: relative;
}

.c0.primary {
  width: 200px;
  padding: 8px;
  border-radius: 8px;
}

.c0.secondary {
  width: 200px;
  padding: 8px;
  border-radius: 8px 0;
}

.c0.error {
  border: 1px solid #ff0000;
  color: #ff0000;
}

.c0.error:focus {
  border: 2px solid #ff0000;
}

.c0 select {
  display: none;
}

.c0.custom-select {
  position: relative;
  display: inline-block;
  width: 100%;
}

.c0 .custom-select-trigger {
  width: 100%;
  cursor: pointer;
  height: 20px;
}

.c0 .custom-options {
  background-color: '';
  display: none;
  position: absolute;
  top: 100%;
  left: 0px;
  right: 0;
  background-color: #fff;
  border: 1px solid #00abff;
  max-width: 215px;
  width: 100%;
  max-height: 114px;
  overflow-y: auto;
  z-index: 1;
}

.c0 .custom-options.primary {
  border-radius: 8px;
}

.c0 .custom-options.secondary {
  border-radius: 8px 0;
}

.c0 .custom-option {
  padding: 10px;
  cursor: pointer;
}

.c0 .custom-option:hover {
  background-color: '';
  color: #00abff;
}

<div>
   
  <div
    class="c0 primary undefined"
  >
    <div
      class="custom-select"
    >
      <label
        class="c1"
      >
         
      </label>
      <div
        class="custom-select-trigger primary undefined"
        data-testid="select-undefined"
      />
      <div
        class="custom-options primary undefined"
      >
        <div
          class="custom-option"
          data-testid="options"
          data-value="Option 1"
        >
          Option 1
        </div>
        <div
          class="custom-option"
          data-testid="options"
          data-value="Option 2"
        >
          Option 2
        </div>
        <div
          class="custom-option"
          data-testid="options"
          data-value="Option 3"
        >
          Option 3
        </div>
      </div>
      <select
        id="real-select"
      >
        <option
          value="Option 1"
        >
          Option 1
        </option>
        <option
          value="Option 2"
        >
          Option 2
        </option>
        <option
          value="Option 3"
        >
          Option 3
        </option>
      </select>
    </div>
  </div>
   
</div>
`;
